bin/l10n-sync.js

```diff
@@ -0,0 +1,352 @@
+  // count for safety.
+  const totalTryCount = 3;
+  let error /*: Error | null */ = null;
+  let tryCount = 0;
+
+  // Try to sync and retry for `totalTryCount` times if it fails.
+  do {
+    try {
+      if (tryCount > 0) {
+        console.warn(stripIndent`
+        Syncing the l10n branch has failed.
+        This may be due to a new commit during this operation. Trying again.
+        Tried ${fewTimes(tryCount)} out of ${totalTryCount}.
+      `);
+      }
+
+      console.log(`>>> Fetching upstream ${upstream}.`);
+      await logAndExec('git', 'fetch', upstream);
+
+      // First, check if the l10n branch contains only changes in `locales` directory.
+      await checkAllowedPaths({
+        upstream,
+        compareBranch: 'l10n',
+        baseBranch: 'main',
+        matchRegexp: vendoredLocalesPath,
+      });
+
+      // Second, check if the main branch contains changes except the translated locales.
+      await checkAllowedPaths({
+        upstream,
+        compareBranch: 'main',
+        baseBranch: 'l10n',
+        matchRegexp: nonVendoredLocalesPath,
+      });
+
+      console.log('>>> Merging main to l10n.');
+      await logAndExec('git', 'checkout', `${upstream}/l10n`);
+
+      const currentDate = DATE_FORMAT.format(new Date());
+      await logAndExec(
+        'git',
+        'merge',
+        `${upstream}/main`,
+        '-m',
+        `🔃 Daily sync: main -> l10n (${currentDate})`
+      );
+
+      console.log(`>>> Pushing to ${upstream}'s l10n branch.`);
+      await pauseWithMessageIfNecessary();
+      const pushResult = await logAndExec(
+        'git',
+        'push',
+        '--no-verify',
+        upstream,
+        'HEAD:l10n'
+      );
+
+      // Print the output of `git push` so user can see the result.
+      console.log(pushResult.stdout.toString());
+      console.log(pushResult.stderr.toString());
+
+      // Clear out the error after everything is done, in case this is a retry.
+      error = null;
+    } catch (e) {
+      error = e;
+    }
+    tryCount++;
+  } while (error !== null && tryCount < totalTryCount);
+
+  if (error) {
+    console.error(
+      `Tried to sync the l10n branch ${fewTimes(totalTryCount)} but failed.`
+    );
+    throw error;
+  }
+}
+
+/**
+ * Main function to be executed in the global scope.
+ *
+ * @throws Will throw an error if any of the functions it calls throw.
+ */
+async function main() /*: Promise<void> */ {
+  const args = process.argv.slice(2);
+
+  if (args.includes('-y')) {
+    SKIP_PROMPTS = true;
+  }
+
+  const upstream = await findUpstream();
+  await checkIfWorkspaceClean();
+
+  console.log(
+    `This script will sync the l10n branch on the remote '${upstream}' with the main branch.`
+  );
+  await pauseWithMessageIfNecessary('Are you sure?');
+
+  await tryToSync(upstream);
+
+  console.log('>>> Going back to your previous banch.');
+  await logAndExec('git', 'checkout', '-');
```

Comment:
I would do this at the end of `tryToSync`, because that's where we `git checkout` the l10n branch too.

Reply:
Yeah, that makes sense. Moved it there.